tests/unit/utils_test/test_core_GenericDataChunkIterator.py

```diff
@@ -0,0 +1,175 @@
+
+    def check_direct_hdf5_write(self, iterator_options):
+        iterator = self.TestNumpyArrayDataChunkIterator(array=self.test_array, **iterator_options)
+        with h5py.File(name=self.test_dir / "test_generic_iterator_array.hdf5", mode="w") as f:
+            dset = f.create_dataset(
+                name="test", shape=self.test_array.shape, dtype="int16", chunks=iterator.chunk_shape
+            )
+            for chunk in iterator:
+                dset[chunk.selection] = chunk.data
+            np.testing.assert_array_equal(np.array(dset), self.test_array)
+            self.assertEqual(dset.chunks, iterator.chunk_shape)
+
+    def test_abstract_assertions(self):
+        class TestGenericDataChunkIterator(GenericDataChunkIterator):
+            pass
+        with self.assertRaises(TypeError) as error:
+            TestGenericDataChunkIterator()
+        self.assertEqual(
+            str(error.exception),
+            "Can't instantiate abstract class TestGenericDataChunkIterator with abstract methods "
+            "_get_data, _get_dtype, _get_maxshape"
+        )
+
+    def test_joint_option_assertions(self):
+        with self.assertRaises(AssertionError) as error:
+            self.TestNumpyArrayDataChunkIterator(array=self.test_array, buffer_shape=(2000, 384), buffer_gb=1)
+        self.assertEqual(str(error.exception), "Only one of 'buffer_gb' or 'buffer_shape' can be specified!")
+
+        with self.assertRaises(AssertionError) as error:
+            self.TestNumpyArrayDataChunkIterator(array=self.test_array, chunk_shape=(1580, 316), chunk_mb=1)
+        self.assertEqual(str(error.exception), "Only one of 'chunk_mb' or 'chunk_shape' can be specified!")
+
+        buffer_shape = (1000, 192)
+        chunk_shape = (100, 384)
+        with self.assertRaises(AssertionError) as error:
+            self.TestNumpyArrayDataChunkIterator(array=self.test_array, buffer_shape=buffer_shape, chunk_shape=chunk_shape)
+        self.assertEqual(
+            str(error.exception),
+            f"Some dimensions of chunk_shape ({chunk_shape}) exceed the manual buffer shape ({buffer_shape})!"
+        )
+
+        buffer_shape = (1000, 192)
+        chunk_shape = (1000, 5)
+        with self.assertRaises(AssertionError) as error:
+            self.TestNumpyArrayDataChunkIterator(array=self.test_array, buffer_shape=buffer_shape, chunk_shape=chunk_shape)
+        self.assertEqual(
+            str(error.exception),
+            f"Some dimensions of chunk_shape ({chunk_shape}) do not evenly divide the buffer shape ({buffer_shape})!"
+        )
+
+    def test_buffer_option_assertions(self):
+        buffer_gb = -1
+        with self.assertRaises(AssertionError) as error:
+            self.TestNumpyArrayDataChunkIterator(array=self.test_array, buffer_gb=buffer_gb)
+        self.assertEqual(str(error.exception), f"buffer_gb ({buffer_gb}) must be greater than zero!")
```

Comment:
```suggestion
        with self.assertRaises(AssertionError, msg=f"buffer_gb (-1) must be greater than zero!") as error:
            self.TestNumpyArrayDataChunkIterator(array=self.test_array, buffer_gb=buffer_gb)
```

Reply:
`assertRaises(error, msg=msg)` doesn't check that the error message equals `msg`. Instead, it outputs `msg` on error. https://stackoverflow.com/questions/58687475/self-assertraises-as-a-context-manager-with-msg-argument-not-working-as-expected

Use [`self.assertRaisesWith`](https://hdmf.readthedocs.io/en/stable/hdmf.testing.testcase.html#hdmf.testing.testcase.TestCase.assertRaisesWith) for an exact match, or `self.assertRaisesRegex` for a regex match instead.

Reply:
To clean up the test code, I recommend changing all of the message tests here to use `self.assertRaisesWith`.

Reply:
Thanks @rly, didn't know about that function but it's really nice! Swapped all assertion tests to this format.